src/main/java/mate/academy/security/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,45 @@
+package mate.academy.security;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> user = userService.findByEmail(email);
+        if (user.isPresent()) {
+            User existingUser = user.get();
+            byte[] salt = existingUser.getSalt();
+            String hashedPassword = HashUtil.hashPassword(password, salt);
+            if (hashedPassword.equals(existingUser.getPassword())) {
+                return existingUser;
+            }
+        }
+        throw new AuthenticationException("Incorrect email or password");
+    }
+
+    @Override
+    public User register(String email, String password) throws RegistrationException {
+        if (userService.findByEmail(email).isPresent()) {
+            throw new RegistrationException("Email is already in use");
```

Comment:
Similarly, consider logging the registration attempt with an already used email before throwing the `RegistrationException`.